---
title: 简单的Promise
date: 2023-12-30
category: JavaScript
---

## 什么是 promise

`Promise` 是异步编程的一种解决方案 比传统的解决方案——回调函数和事件——更合理和更强大 它由社区最早提出和实现 `ES6` 将其写进了语言标准 统一了用法 原生提供了 `Promise` 对象。

## promise 状态

`promise` 对象代表一个异步操作 有三种状态 `pending`(进行中)、`fulfilled`(已成功)、`reject`(已失败) 只有 `resolve` 和 `reject` 能改变 `promise` 的状态

## promise 的使用方法

```js
const p = new Promise((resolve, reject) => {
	if ('成功了') {
		resolve('成功')
	} else {
		reject('失败')
	}
})
```

`Promise`构造函数接受一个函数作为参数 该函数的两个参数分别是`resolve`和`reject` 由`javaScript`提供
`resolve`
此函数将 `promise` 的状态从 未完成 -> 已完成(pending -> fulfilled)
`reject`
此函数将 `promise` 的状态从 未完成 -> 已失败(pending -> reject)

## 消费 promise

```js
p.then(res => {
    // promise 成功 走这里
    console.log(res)
})
.catch(err) {
    // 出错走这里
    console.log(err)
}
```

## 提问环节

问：`promise.then` 是异步还是同步

> 答：异步 `promise` 本身是同步的 但是会等到结果被 `resolve` 才会走到 `then` 方法

## 参考资料

关于 promise 更多的使用方式 请查看参考资料

<a href="https://es6.ruanyifeng.com/#docs/promise" target="_blank">
	ECMAScript 6 - promise
</a>

## 手写简单的Promise

```js
const PENDING = 'PENDING'
const FULFILLED = 'FULFILLED'
const REJECTED = 'REJECTED'

class Promise {
	constructor(executor) {
		// 默认状态 PENDING
		this.status = PENDING
		// 成功的值
		this.value = undefined
		// 失败原因
		this.reason = undefined

		// 存放成功的回调函数
		this.onResolvedCallbacks = []
		// 存放失败的回调函数
		this.onRejectedCallbacks = []

		const resolve = value => {
			if (this.status === PENDING) {
				this.status = FULFILLED
				this.value = value
				// 调用存好的回调函数
				this.onResolvedCallbacks.forEach(fn => fn())
			}
		}

		const reject = reason => {
			if (this.status === PENDING) {
				this.status = REJECTED
				this.reason = reason

				this.onRejectedCallbacks.forEach(fn => fn())
			}
		}

		try {
			executor(resolve, reject)
		} catch (error) {
			reject(error)
		}
	}

	then(onFulfilled, onRejected) {
		if (this.status === FULFILLED) {
			onFulfilled(this.value)
		}

		if (this.status === REJECTED) {
			onRejected(this.reason)
		}

		// 当调用 then 方法时 状态还是pending 说明是异步的操作 先存好回调函数
		// 在resolve或者是reject触发时去调用
		if (this.status == PENDING) {
			this.onResolvedCallbacks.push(() => {
				onFulfilled(this.value)
			})
			this.onRejectedCallbacks.push(() => {
				onRejected(this.value)
			})
		}
	}
}
```

测试

```js
const promise = new Promise((resolve, reject) => {
	// resolve('同步成功')
	setTimeout(() => {
		resolve('异步成功')
	}, 1000)
})
promise.then(
	data => {
		console.log('success', data)
	},
	err => {
		console.log('fail', err)
	},
)
```

!!! note TIPS
手写的 Promise 在直接使用 resolve 的时候(不使用异步的方法时)是同步的 和实际有差别 感兴趣可以自行研究 这里只是实现了基础的功能
!!!
